controllers/clustercache/cluster_cache.go

```diff
@@ -0,0 +1,650 @@
+			return ctrl.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		log.Error(err, "Failed to read the Cluster object")
+		return ctrl.Result{RequeueAfter: defaultRequeueAfter}, nil
+	}
+
+	// Return if infrastructure is not ready yet to avoid trying to open a connection when it cannot succeed.
+	// Requeue is not needed as there will be a new reconcile.Request when Cluster.status.infrastructureReady is set.
+	if !cluster.Status.InfrastructureReady {
+		log.V(4).Info("Infrastructure is not ready yet")
+		return reconcile.Result{}, nil
+	}
+
+	// Track if the current Reconcile is doing a connect or disconnect.
+	didConnect := false
+	didDisconnect := false
+
+	requeueAfterDurations := []time.Duration{}
+
+	// Try to connect, if not connected.
+	connected := accessor.Connected(ctx)
+	if !connected {
+		lastClientCreationFailureTimestamp := accessor.GetLastClientCreationFailureTimestamp(ctx)
+
+		// Requeue, if client creation failed within the Client.CreationRetryInterval.
+		if requeueAfter, requeue := shouldRequeue(time.Now(), lastClientCreationFailureTimestamp, accessor.config.Client.CreationRetryInterval); requeue {
+			requeueAfterDurations = append(requeueAfterDurations, requeueAfter)
+		} else {
+			// Try to connect
+			if err := accessor.Connect(ctx); err != nil {
+				// Requeue, if the connect failed.
+				log.Error(err, "Failed to connect cluster accessor")
+
+				lastClientCreationFailureTimestamp = time.Now()
+				accessor.SetLastClientCreationFailureTimestamp(ctx, lastClientCreationFailureTimestamp)
+				if requeueAfter, requeue := shouldRequeue(time.Now(), lastClientCreationFailureTimestamp, accessor.config.Client.CreationRetryInterval); requeue {
+					requeueAfterDurations = append(requeueAfterDurations, requeueAfter)
+				}
+			} else {
+				// Store that connect was done successfully.
+				didConnect = true
+				connected = true
+			}
+		}
+	}
+
+	// Run the health probe, if connected.
+	if connected {
+		lastProbeTimestamp := accessor.GetLastProbeTimestamp(ctx)
+
+		// Requeue, if health probe was already run within the HealthProbe.Interval.
+		if requeueAfter, requeue := shouldRequeue(time.Now(), lastProbeTimestamp, accessor.config.HealthProbe.Interval); requeue {
+			requeueAfterDurations = append(requeueAfterDurations, requeueAfter)
+		} else {
+			// Run the health probe
+			lastProbeTimestamp, healthCheckFailed := accessor.HealthCheck(ctx)
+			if healthCheckFailed {
+				// Disconnect if the health probe failed (either with unauthorized or consecutive failures >= HealthProbe.FailureThreshold).
+				accessor.Disconnect(ctx)
+
+				// Store that disconnect was done.
+				didDisconnect = true
+				connected = false //nolint:ineffassign // connected is *currently* not used below, let's make update it anyway
+			}
+
+			// Requeue for next health probe.
+			if requeueAfter, requeue := shouldRequeue(time.Now(), lastProbeTimestamp, accessor.config.HealthProbe.Interval); requeue {
```

Comment:
I would not do it.

The problem is, what if the kubeconfig in the Secret is broken and stays broken. Requeuing immediately means we will get an infinite Reconcile with 0 delay.

We could technically implement the connect directly here, but I'm not sure if the time we safe is worth making this more complex. 

Also because the requeue here is just 10s, so we would safe 10s at most

Reply:
Okay found a way to use this to simplify in other places